nengo_loihi/splitter.py

```diff
@@ -1,543 +1,141 @@
-        networks.needs_sender[conn.learning_rule] = PESModulatoryTarget(probe)
-    networks.remove(conn)
-
-
-def split_host_to_learning_rules(networks, conns):
-    for conn in conns:
-        if conn in networks:
-            # Already processed
-            continue
-
-        pre_loc = networks.location(conn.pre_obj)
-        if (pre_loc == "host"
-                and isinstance(conn.post_obj, LearningRule)):
-            split_host_to_learning_rule(networks, conn)
-            assert conn in networks
-
-
-def split_host_to_learning_rule(networks, conn):
-    dim = conn.size_out
-    logger.debug("Creating HostSendNode for %s", conn)
-    send = HostSendNode(
-        dim,
-        label=None if conn.label is None else "%s_send" % conn.label,
-        add_to_container=False,
-    )
-    networks.add(send, "host")
-
-    pre2send = Connection(
-        conn.pre,
-        send,
-        function=conn.function,
-        solver=conn.solver,
-        eval_points=conn.eval_points,
-        scale_eval_points=conn.scale_eval_points,
-        synapse=conn.synapse,
-        transform=conn.transform,
-        label=conn.label,
-        add_to_container=False,
-    )
-    networks.add(pre2send, "host")
-    pes_target = networks.needs_sender[conn.post_obj]
-    networks.host2chip_senders[send] = pes_target
-    networks.remove(conn)
-
-
-def place_probes(networks):
-    for probe in networks.original.all_probes:
-        target = base_obj(probe.target)
-        networks.move(probe, networks.location(target))
-
-
-def split_pre_from_host(networks):  # noqa: C901
-    logger.info("Splitting pre model from host")
-
-    inputs = defaultdict(list)
-    outputs = defaultdict(list)
-    queue = []
-
-    for d in [networks.moves, networks.adds]:
-        for obj in d:
-            if isinstance(obj, Connection):
-                inputs[base_obj(obj.post_obj)].append(obj)
-                outputs[base_obj(obj.pre_obj)].append(obj)
-            elif isinstance(obj, HostSendNode):
-                networks.move(obj, "host_pre", force=True)
+        def mark_precomputable(obj):
+            assert isinstance(obj, (Node, Ensemble))
+            if obj not in precomputable:
+                precomputable.add(obj)
                 queue.append(obj)
 
-    while len(queue) > 0:
-        node_or_ens = queue.pop()
-
-        for conn in inputs[node_or_ens] + outputs[node_or_ens]:
-            if networks.location(conn) != "host":
-                continue
-            networks.move(conn, "host_pre", force=True)
-
-            if conn in inputs[node_or_ens]:
-                obj = base_obj(conn.pre_obj)
-            elif conn in outputs[node_or_ens]:
-                obj = base_obj(conn.post_obj)
-
-            if (isinstance(obj, (Node, Ensemble))
-                    and networks.location(obj) == "host"):
-                if isinstance(obj, HostReceiveNode):
+        # initialize queue with the pre objects on
+        # host -> chip connections
```

Comment:
Why are all `pre` objects on host->chip connections precomputable? Couldn't you have something like node->ensemble->node->ensemble->node (where nodes are on host and ensembles are on chip) where the middle node would be the `pre` of a host->chip connection but not be precomputable?

Reply:
Great catch... I'm amazed that none of the tests caught this. I pushed a commit 0cbe29b which tests this and makes the fix (namely, using the right set of connections to form the adjacency lists).

Reply:
This still appears to assume that all `pre` objects on host->chip connections are precomputable. This might be true in some cases, like in the test you added where the passthrough nodes can be removed, but what if those weren't passthroughs (i.e. if all nodes in that test computed functions)? I think in general, you can have networks where some `pre` objects of connections are not precomputable.

Reply:
Yep.. this is the same assumption that passthrough removal has always been making. If anything in the transitive closure of objects that send to host (after passthrough removal) also receives data from the host then a `BuildError` is raised. That is, the example that you suggested currently (and previously) raises an error. There is a unit test for this (`test_split_precompute_loop_error`). Note I didn't change the algorithm here. I just updated it to be a bit clearer IMO and more efficient (it used to be `O(n^2)`). I talked about this in slack but there was no interest in making this smarter at the moment. I've summarized this alongside a similar issue now in #214.

The difference that you initially pointed out stemmed from the fact that previously the passthrough removal would be physically applied to the `SplitNetworks` object before computing the closure. In this PR the list of connections weren't changed, so I added the above commit to make sure it's using the same list as it had been in the original splitter.

By the way we should remove "*Various improvements to passthrough removal, including not removing useful passthrough nodes.*" from the commit message. This was squashed from one of the other commits that was fixing a mistake I had previously made. There aren't any improvements to passthrough removal in this PR apart from making it `O(n)` instead of `O(n^2)`.

Reply:
Okay, maybe some comments in the code to that effect would be helpful. I can draft them.

That's good to know about that part of the commit message. I had wondered about that when I put it in, because I couldn't see where the stuff about "not removing useful passthrough nodes" happened, but I didn't want to get rid of it until I checked with you. As for "various improvements", I was referring to renaming "offchip" to "ignore", supporting `ObjView`, and returning a named tuple instead of a tuple. I don't actually see that the passthrough removal algorithm changes at all.

Reply:
"I don't actually see that the passthrough removal algorithm changes at all." Which part are you referring to here? The `O(n^2) -> O(n)` part? This is just from avoiding the repeated list pops, which each take `O(n)` time, by instead advancing a `head` index.

Reply:
List pops in Python are O(1) (see [time complexity](https://wiki.python.org/moin/TimeComplexity))

Reply:
That's from an amortized analysis or average case analysis. Here we were always popping from the beginning of a list of size `O(n)` which is `O(n)` each pop. Quote from the same page:

> Internally, a list is represented as an array; the largest costs come from growing beyond the current allocation size (because everything must move), or from inserting or deleting somewhere near the beginning (because everything after that must move). If you need to add/remove at both ends, consider using a collections.deque instead.

Since lists are just arrays, `list.pop(0)` needs to shift the entire array each time.

Reply:
I thought we were doing `queue.pop()` ? Which pops from the end of the list

Reply:
Ah, you are right. The naming of the variable was a misnomer. Since we append to the last and pop from the last, it is not a queue -- it is a _stack_. It doesn't actually matter which one it is, but I was trying to keep things consistent with a queue. So either we can change it back to `pop` and rename it from `queue -> stack` or leave it as is. The ordering shouldn't matter, but should make it clear.

Reply:
This isn't the passthrough removal algorithm, though, this is the splitter. That's all I was trying to say, is that I didn't see any changes to passthrough removal.

Reply:
Apparently my brain has started interchanging the words "passthrough removal" and "splitter" as well as "chip" and "host". Ha ha .. :x You're right that I didn't change anything about passthrough removal other than fixing a bug that affected probed `ObjView`s.

Reply:
Pushed commit 433b489 and updated issue #214 according to the above conversation about precomputing.